File: py/graph/graph.py
from queue_graphs import Queue

class Vertex:
    def __init__(self, value):
        self.name = value
        self.neighbors = []

    def add_neighbor(self, vertex):
        if vertex not in self.neighbors:
            self.neighbors.append(vertex)
            self.neighbors.sort()


class Graph:
    def __init__(self):
        self.vertices = {}

    def add_vertex(self, vertex):
        new_vertex = Vertex(vertex)
        if new_vertex.name not in self.vertices:
            self.vertices[new_vertex.name] = new_vertex
            return True
        else:
            return False
    
    def add_edge(self, src, dest):
        if src in self.vertices and dest in self.vertices:
            for key, value in self.vertices.items():
                if key == src:
                    value.add_neighbor(dest)
                if key == dest:
                    value.add_neighbor(src)
            return True
        else: 
            return False
    
    def print_graph(self):
        for vert in self.vertices.keys():
            print(vert + " " + str(self.vertices[vert].neighbors))


    # def breadth_first_search(self, vert):
    #     bfs_q, visited = Queue(), {}
    #     bfs_q.enqueue(vert)

    #     while not bfs_q.is_empty():
    #         crntQItem = bfs_q.dequeue()
    #         print("Vertex ", crntQItem.value)

    #         vrtxConns = self.vertices[crntQItem.value].neighbors
    #         print("Connections ", vrtxConns)

    #         for item in vrtxConns:
    #             if visited[item] != True:
    #                 bfs_q.enqueue(item)
    #                 visited[item] = True

File: ts/src/graph/graph.ts
import Queue from '../queue/queue'

class Graph {
    ajdacentList: any
    verticesList: any

    constructor() {
        this.ajdacentList = {}
        this.verticesList = []
    }

    addVertex(vertex: any) {
        if (this.verticesList.indexOf(vertex) < 0) {
            this.verticesList.push(vertex)
            this.ajdacentList[vertex] = []
        } else {
            return "Vertex exists Already"
        }
    }

    addEdge(src: any, dest: any) {
        this.ajdacentList[src].push(dest)
        this.ajdacentList[dest].push(src)
    }

    printGraph() {
        var vertices = this.verticesList
        for (var i in vertices) {
            var vrtx = vertices[i]
            var edges = this.ajdacentList[vrtx]
            console.log("Vertex ", vrtx, "Connections ", edges)
        }
    }

    breadthFirstSearch(vertex: any) {
        var visited: any = []
        var q = new Queue()

        q.enqueue(vertex)

        while (!q.isEmpty()) {
            var currentQueueItem = q.dequeue()
            console.log("Vertex ", currentQueueItem)

            var vertexConns = this.ajdacentList[currentQueueItem]
            console.log("Connections of", currentQueueItem, vertexConns)

            for (var i = 0; i < vertexConns.length; i++) {
                var item = vertexConns[i]

                if (!visited[item] == true) {
                    visited[item] = true
                    // console.log(item, " not visited, marking as visited now")
                    q.enqueue(item)
                } else {
                    // console.log(item, "visited already")
                }
            }
        }
    }

    depthFirstSearch(vertex: any) {
        var visited: any = []
        this.DSFChecker(vertex, visited)
    }

    private DSFChecker(vert: any, visited: any) {
        visited[vert] = true
        console.log("Vertex ", vert)

        var vertConns = this.ajdacentList[vert]
        console.log("Connections ", vertConns)

        for (var i = 0; i < vertConns.length; i++) {
            var item = vertConns[i]
            if (visited[item] != true) {
                this.DSFChecker(item, visited)
            }
        }
    }

}
export default Graph